pkg/distributor/query.go

```diff
@@ -232,22 +255,53 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 					}
 				}
 
-				if chunkBytesLimitErr := queryLimiter.AddChunkBytes(resp.ChunksSize()); chunkBytesLimitErr != nil {
+				if chunkBytesLimitErr := queryLimiter.AddChunkBytes(ingester_client.ChunksSize(resp.Chunkseries)); chunkBytesLimitErr != nil {
 					return ingesterQueryResult{}, validation.LimitError(chunkBytesLimitErr.Error())
 				}
 
 				result.chunkseriesBatches = append(result.chunkseriesBatches, resp.Chunkseries)
+			} else if len(resp.Series) > 0 {
+				labelsBatch := make([]labels.Labels, 0, len(resp.Series))
+				streamingSeriesCount += len(resp.Series)
+
+				for _, s := range resp.Series {
+					if limitErr := queryLimiter.AddSeries(s.Labels); limitErr != nil {
+						return ingesterQueryResult{}, validation.LimitError(limitErr.Error())
+					}
+
+					labelsBatch = append(labelsBatch, mimirpb.FromLabelAdaptersToLabels(s.Labels))
+				}
+
+				streamingSeriesBatches = append(streamingSeriesBatches, labelsBatch)
+			}
+
+			if resp.IsEndOfSeriesStream {
+				if streamingSeriesCount > 0 {
+					result.streamingSeries.Series = make([]labels.Labels, 0, streamingSeriesCount)
+
+					for _, batch := range streamingSeriesBatches {
+						result.streamingSeries.Series = append(result.streamingSeries.Series, batch...)
+					}
+
+					streamReader := ingester_client.NewSeriesChunksStreamReader(stream, streamingSeriesCount, queryLimiter)
+					closeStream = false
+					result.streamingSeries.StreamReader = streamReader
+				}
+
+				return result, nil
 			}
 		}
 	}
 
 	cleanup := func(result ingesterQueryResult) {
-		// Nothing to do.
+		if result.streamingSeries.StreamReader != nil {
+			result.streamingSeries.StreamReader.Close()
+		}
 	}
 
 	results, err := ring.DoUntilQuorum(ctx, replicationSet, queryIngester, cleanup)
 	if err != nil {
-		return nil, err
+		return ingester_client.CombinedQueryStreamResponse{}, err
 	}
 
 	// We keep track of the number of chunks that were able to be deduplicated entirely
```

Comment:
[nit] I would improve this comment saying that, in the case of streaming series, the deduplication is done elsewhere.

Reply:
Good point, added in e434f8e.